data/tokenizer.py

```diff
@@ -120,3 +126,52 @@ def untokenize(self, tokens: list[str]) -> pd.DataFrame:
         df = pd.DataFrame(velocities, columns=["velocity"])
 
         return df
+
+
+class DstartEncoder(MidiEncoder):
+    def __init__(self, bins: int = 200):
+        self.specials = ["<s>", "</s>", "<blank>"]
+        self.bins = bins
+        # Make a copy of special tokens ...
+        self.vocab = list(self.specials)
+        self.bin_edges = self.load_bin_edges()
+        # ... and add velocity tokens
+        self._build_vocab()
+        self.token_to_id = {token: it for it, token in enumerate(self.vocab)}
+
+    def load_bin_edges(self):
+        artifacts_path = to_absolute_path("artifacts/bin_edges.yaml")
+        with open(artifacts_path, "r") as f:
+            bin_edges = yaml.safe_load(f)
+
+        dstart_bin_edges = bin_edges["dstart"][self.bins]
+        return dstart_bin_edges
+
+    def _build_vocab(self):
+        self.vocab += [str(possible_bin) for possible_bin in range(self.bins)]
+
+    @property
+    def vocab_size(self) -> int:
+        return len(self.vocab)
+
+    def tokenize(self, record: dict) -> list[str]:
+        # TODO I don't love the idea of adding tokens durint *tokenize* call
+        # If we want to pretend that our midi sequences have start and finish
+        # we should take care of that before we get here :alarm:
+        dstart = []
+        for it in range(len(record["start"]) - 1):
+            dstart.append(record["start"][it + 1] - record["start"][it])
+        dstart.append(0)
+
+        dstart_bins = np.digitize(dstart, self.bin_edges) - 1
+
+        # get tokens from quantized data
+        tokens = [str(dstart_bin) for dstart_bin in dstart_bins]
+        tokens = ["<s>"] + tokens + ["</s>"]
+        return tokens
```

Comment:
I've managed to speed up the quantization during tokenization process 31 times on average by omitting converstion to pd.DataFrame and operating on dict instead.